File: mobile/ios/Sources/Transport/HostAllowlist.swift
import Foundation

public enum HostAllowlist {
    // Cleartext is fine over WireGuard and on a home LAN, and nowhere else.
    // This is the whole reason the app can ship without TLS setup, so it is
    // enforced in one place and tested directly.
    public static func validate(_ url: URL) throws -> URL {
        guard let scheme = url.scheme?.lowercased(), let host = url.host?.lowercased() else {
            throw DaemonError.unreachableHost(url.absoluteString)
        }
        // A phone can never reach the daemon's own loopback address, over any
        // scheme — refused unconditionally, per mobile/PROTOCOL.md §10. This must
        // run before the https branch below, or https://127.0.0.1 would sail
        // through on "https is always allowed" alone.
        if isLoopback(host) { throw DaemonError.unreachableHost(host) }
        if scheme == "https" { return url }
        guard scheme == "http" else { throw DaemonError.unreachableHost(url.absoluteString) }
        if host.hasSuffix(".ts.net") { return url }
        if isPrivateOrTailscale(host) { return url }
        throw DaemonError.unreachableHost(host)
    }

    private static func isLoopback(_ host: String) -> Bool {
        host == "127.0.0.1" || host == "localhost" || host == "::1" || host == "[::1]"
    }

    private static func isPrivateOrTailscale(_ host: String) -> Bool {
        let parts = host.split(separator: ".").compactMap { Int($0) }
        guard parts.count == 4, parts.allSatisfy({ (0...255).contains($0) }) else { return false }
        switch (parts[0], parts[1]) {
        case (10, _): return true
        case (192, 168): return true
        case (172, 16...31): return true
        case (100, 64...127): return true   // Tailscale CGNAT range
        default: return false
        }
    }
}

File: mobile/android/app/src/main/kotlin/sh/openagi/mobile/transport/HostAllowlist.kt
package sh.openagi.mobile.transport

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

// Cleartext is fine over WireGuard and on a home LAN, and nowhere else.
// This is the whole reason the app can ship without TLS setup, so it is
// enforced in one place and tested directly. Keep this table byte-for-byte
// in agreement with mobile/ios/Sources/Transport/HostAllowlist.swift.
object HostAllowlist {
    fun validate(raw: String): HttpUrl {
        val url = raw.toHttpUrlOrNull() ?: throw DaemonException.UnreachableHost(raw)
        // Drop any path so callers can append their own without doubling it.
        val origin = HttpUrl.Builder()
            .scheme(url.scheme)
            .host(url.host)
            .port(url.port)
            .build()
        // Loopback is refused whatever the scheme, per PROTOCOL.md §10: a phone
        // cannot reach its own loopback, so accepting one turns a pairing typo
        // into a silent hang instead of an immediate, legible refusal. This check
        // precedes the scheme branch deliberately — putting it after would let
        // https://127.0.0.1 through, which is the gap iOS shipped and had to fix.
        if (url.host.lowercase() in setOf("127.0.0.1", "localhost", "::1")) {
            throw DaemonException.UnreachableHost(url.host)
        }
        if (url.scheme == "https") return origin
        if (url.scheme != "http") throw DaemonException.UnreachableHost(raw)
        val host = url.host.lowercase()
        if (host.endsWith(".ts.net")) return origin
        if (isPrivateOrTailscale(host)) return origin
        throw DaemonException.UnreachableHost(host)
    }

    private fun isPrivateOrTailscale(host: String): Boolean {
        val parts = host.split(".").mapNotNull { it.toIntOrNull() }
        if (parts.size != 4 || parts.any { it !in 0..255 }) return false
        val (a, b) = parts
        return when {
            a == 10 -> true
            a == 192 && b == 168 -> true
            a == 172 && b in 16..31 -> true
            a == 100 && b in 64..127 -> true // Tailscale CGNAT range
            else -> false
        }
    }
}
